nnfabrik/templates/transfer/transferred_trained_model.py

```diff
@@ -0,0 +1,292 @@
+from nnfabrik.utility.dj_helpers import gitlog, make_hash
+from nnfabrik.templates.trained_model import TrainedModelBase
+
+
+class TransferredTrainedModelBase(TrainedModelBase):
+    """
+    A modified version of TrainedModel table which enables step-wise table population given the
+    step specification. Refer to the corresponding example notebook for a demo.
+    """
+
+    table_comment = "Transferred trained models"
+
+    @property
+    def definition(self):
+        definition = """
+        # {table_comment}
+        transfer_step:                     int          # transfer step
+        -> self().model_table
+        -> self().dataset_table
+        -> self().trainer_table
+        -> self().seed_table
+        collapsed_history:                 varchar(64)
+        data_transfer:                     bool 
+        ---
+        comment='':                        varchar(768) # short description 
+        score:                             float        # loss
+        output:                            longblob     # trainer object's output
+        ->[nullable] self().user_table
+        trainedmodel_ts=CURRENT_TIMESTAMP: timestamp    # UTZ timestamp at time of insertion
+        """.format(
+            table_comment=self.table_comment
+        )
+        return definition
+
+    class ModelStorage(TrainedModelBase.ModelStorage):
+        pass
+
+    class DataStorage(dj.Part):
+        @property
+        def definition(self):
+            definition = """
+            # Contains the data generated by the transfer step, stored externally.
+            -> master
+            ---
+            transfer_data:            attach@{storage}
+            """.format(
+                storage=self._master.storage
+            )
+            return definition
+
+    class CollapsedHistory(dj.Part):
+        """
+        For the result of two or more transfer steps to be uniquely identifiable,
+        we compress its entire history (the keys of all previous steps) into a single hash (`collapsed_history`).
+        This table keeps track of this process and can be used to recursively retrieve the transfer history.
+        """
+
+        definition = """
+        next_collapsed_history:                 varchar(64)
+        -> master
+        """
+
+        @classmethod
+        def add_entry(cls, key):
+            key = {
+                k: key[k]
+                for k in [
+                    "model_fn",
+                    "model_hash",
+                    "dataset_fn",
+                    "dataset_hash",
+                    "trainer_fn",
+                    "trainer_hash",
+                    "seed",
+                    "data_transfer",
+                    "transfer_step",
+                    "collapsed_history",
+                ]
+            }
+            key["next_collapsed_history"] = make_hash(key)
+            cls.insert1(key, skip_duplicates=True)
+
+    def _transfer_recipe(self, transfer_step):
```

Comment:
The name makes me think this is a property. I suggest making it one or changing the name to something more suited for a method (e.g. `_combine_transfer_recipes`).